File: load-tests/negotiation-load.py
"""
load-tests/negotiation-load.py
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
GridNexus Locust fallback load test.

Provides identical 200-concurrent-agent simulation using Locust when k6
is not available in the CI environment.

ASSUMPTION : Locust is installed via `pip install locust`. The
k6 script is the primary tool; this is a documented fallback. Both produce
p50/p95/p99 latency data.

Usage:
    locust -f load-tests/negotiation-load.py \
           --host http://localhost:3000 \
           --users 200 \
           --spawn-rate 20 \
           --run-time 120s \
           --headless \
           --csv load-tests/results/locust
"""

from __future__ import annotations

import json
import random
import time
from locust import HttpUser, task, between, events
from locust.runners import MasterRunner


class MicrogridAgentUser(HttpUser):
    """
    Simulates a single microgrid agent participating in GridNexus.

    Each VU represents one agent executing the full negotiation loop:
      1. Check broker health (liveness)
      2. Poll oracle signals (grid state)
      3. Fetch analytics (simulate negotiation round state)
      4. Fetch topology (simulate stability awareness)
    """

    wait_time = between(0.1, 0.5)  # Inter-round think time

    def on_start(self) -> None:
        """Called once when a simulated user starts."""
        self.agent_id = f"load-agent-{id(self)}"

    @task(1)
    def health_check(self) -> None:
        """Broker liveness probe – always fast."""
        with self.client.get("/health", catch_response=True) as resp:
            if resp.status_code != 200:
                resp.failure(f"Health check failed: {resp.status_code}")

    @task(3)
    def oracle_signal_poll(self) -> None:
        """Agent polling oracle for grid conditions."""
        with self.client.get("/api/oracle-signals", name="/api/oracle-signals", catch_response=True) as resp:
            if resp.status_code != 200:
                resp.failure(f"Oracle signals failed: {resp.status_code}")
            else:
                resp.success()

    @task(5)
    def negotiation_round(self) -> None:
        """Simulates one negotiation round via analytics endpoint."""
        with self.client.get("/api/analytics", name="/api/analytics [negotiation round]", catch_response=True) as resp:
            if resp.status_code != 200:
                resp.failure(f"Analytics endpoint failed: {resp.status_code}")
            else:
                try:
                    body = resp.json()
                    if "summary" not in body:
                        resp.failure("Response missing 'summary' field")
                    else:
                        resp.success()
                except json.JSONDecodeError:
                    resp.failure("Non-JSON response from /api/analytics")

    @task(4)
    def stability_check(self) -> None:
        """Agent checking planar graph topology (Stability Gate)."""
        with self.client.get("/api/topology", name="/api/topology [stability]", catch_response=True) as resp:
            if resp.status_code != 200:
                resp.failure(f"Topology endpoint failed: {resp.status_code}")
            else:
                try:
                    body = resp.json()
                    if "nodes" not in body or not isinstance(body["nodes"], list):
                        resp.failure("Topology response missing nodes")
                    else:
                        resp.success()
                except json.JSONDecodeError:
                    resp.failure("Non-JSON response from /api/topology")


@events.test_stop.add_listener
def on_test_stop(environment, **kwargs) -> None:
    """Print p50/p95/p99 summary to stdout on completion."""
    stats = environment.stats
    print("\n" + "=" * 60)
    print("GridNexus Load Test – Latency Summary")
    print("=" * 60)
    for name, entry in stats.entries.items():
        p50  = entry.get_response_time_percentile(0.50) or 0
        p95  = entry.get_response_time_percentile(0.95) or 0
        p99  = entry.get_response_time_percentile(0.99) or 0
        print(f"  {name}: p50={p50:.0f}ms  p95={p95:.0f}ms  p99={p99:.0f}ms")
    total = stats.total
    print(f"\nTotal requests: {total.num_requests}")
    print(f"Failures:       {total.num_failures}")
    print(f"Failure rate:   {total.fail_ratio*100:.2f}%")
    p99_total = total.get_response_time_percentile(0.99) or 0
    result = "✅ PASS" if p99_total < 2000 and total.fail_ratio < 0.01 else "❌ FAIL"
    print(f"\np99 overall: {p99_total:.0f}ms  →  {result}")
    print("=" * 60)

File: load-tests/negotiation-load.js
/**
 * load-tests/negotiation-load.js
 * ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 * GridNexus 200-concurrent-agent load test.
 *
 * Simulates 200 concurrent microgrid agents, each executing a full
 * negotiation session (Oracle → Bargaining → Stability check).
 *
 * ASSUMPTION : k6 is the preferred load-test runner because it
 * produces structured JSON summaries with p50/p95/p99 directly from
 * --summary-trend-stats and outputs them in machine-readable format via
 * --out json=results/summary.json. If k6 is unavailable in the CI runner,
 * a Locust fallback script is provided in load-tests/negotiation-load.py.
 *
 * Target:
 *   - p99 negotiation-round latency < 2 000 ms
 *   - Zero HTTP 5xx failures
 *   - Zero DB integrity failures (checked via /api/analytics after run)
 *
 * Usage:
 *   k6 run --env BROKER_URL=http://localhost:3000 \
 *          --summary-trend-stats='p(50),p(95),p(99)' \
 *          --out json=load-tests/results/summary.json \
 *          load-tests/negotiation-load.js
 */

import http from "k6/http";
import { check, sleep } from "k6";
import { Trend, Counter, Rate } from "k6/metrics";

// ── Custom metrics ───────────────────────────────────────────────────────────
const negotiationRoundLatency = new Trend("negotiation_round_latency", true);
const stabilityCheckLatency   = new Trend("stability_check_latency",   true);
const oracleSignalLatency     = new Trend("oracle_signal_latency",     true);
const integrityFailures       = new Counter("integrity_failures");
const sessionSuccessRate      = new Rate("session_success_rate");

// ── Test configuration ───────────────────────────────────────────────────────
const BROKER_URL = __ENV.BROKER_URL || "http://localhost:3000";
const ENGINE_URL = __ENV.ENGINE_URL || "http://localhost:8000";

export const options = {
  // Ramp to 200 VUs (Virtual Users = concurrent agents), hold, then ramp down
  stages: [
    { duration: "20s", target: 50  },   // warm-up
    { duration: "30s", target: 200 },   // ramp to 200
    { duration: "60s", target: 200 },   // sustain 200 concurrent agents
    { duration: "10s", target: 0   },   // cool-down
  ],
  thresholds: {
    // Acceptance criteria from 
    "negotiation_round_latency{scenario:default}": ["p(99)<2000"],
    "stability_check_latency{scenario:default}":   ["p(99)<2000"],
    "http_req_failed":                             ["rate<0.01"],   // < 1% error rate
    "session_success_rate":                        ["rate>0.95"],   // > 95% sessions succeed
    "integrity_failures":                          ["count==0"],
  },
};

// ── Helper: generate a unique agent pair ─────────────────────────────────────
function agentPair(vu) {
  return {
    agentA: `load-agent-a-${vu}`,
    agentB: `load-agent-b-${vu}`,
    microgridA: `load-mg-${vu}-a`,
    microgridB: `load-mg-${vu}-b`,
  };
}

// ── Main VU scenario ─────────────────────────────────────────────────────────
export default function () {
  const { agentA, agentB } = agentPair(__VU);
  const headers = { "Content-Type": "application/json" };

  // ── Step 1: Check broker health ─────────────────────────────────────────
  const healthStart = Date.now();
  const healthRes = http.get(`${BROKER_URL}/health`, { headers });
  check(healthRes, {
    "broker health 200": (r) => r.status === 200,
  });

  // ── Step 2: Poll oracle signal (simulates agents checking grid state) ──
  const oracleStart = Date.now();
  const oracleRes = http.get(`${BROKER_URL}/api/oracle-signals`, { headers });
  oracleSignalLatency.add(Date.now() - oracleStart);
  check(oracleRes, {
    "oracle signals 200": (r) => r.status === 200,
  });

  // ── Step 3: Simulate negotiation round via /api/analytics ─────────────
  // (Full WebSocket negotiation requires a WS VU; we measure the REST APIs
  //  that back each round. For WS timing, see load-tests/ws-negotiate.js)
  const roundStart = Date.now();
  const analyticsRes = http.get(`${BROKER_URL}/api/analytics`, { headers });
  negotiationRoundLatency.add(Date.now() - roundStart);

  const analyticsOk = check(analyticsRes, {
    "analytics 200":          (r) => r.status === 200,
    "analytics has summary":  (r) => {
      try {
        const body = JSON.parse(r.body);
        return body.summary !== undefined;
      } catch {
        return false;
      }
    },
  });

  // ── Step 4: Stability check probe ──────────────────────────────────────
  const stabilityStart = Date.now();
  const topologyRes = http.get(`${BROKER_URL}/api/topology`, { headers });
  stabilityCheckLatency.add(Date.now() - stabilityStart);

  const topologyOk = check(topologyRes, {
    "topology 200":       (r) => r.status === 200,
    "topology has nodes": (r) => {
      try {
        const body = JSON.parse(r.body);
        return Array.isArray(body.nodes) && body.nodes.length > 0;
      } catch {
        return false;
      }
    },
  });

  // ── Step 5: Engine stability-check probe ───────────────────────────────
  const engineStabilityStart = Date.now();
  const engStabilityRes = http.post(
    `${ENGINE_URL}/stability/check`,
    JSON.stringify({ coalition: [agentA, agentB] }),
    { headers }
  );
  // Accept 200 (stable) or 422 (validation error with mock IDs) — both are
  // non-5xx and indicate the engine is alive.
  const engineOk = check(engStabilityRes, {
    "engine stability non-5xx": (r) => r.status < 500,
  });

  // Record overall session success
  sessionSuccessRate.add(analyticsOk && topologyOk ? 1 : 0);

  // ── Step 6: Integrity self-check (every 10th VU iteration) ────────────
  if (__ITER % 10 === 0) {
    const intRes = http.get(`${BROKER_URL}/api/analytics`, { headers });
    if (intRes.status !== 200) {
      integrityFailures.add(1);
    }
  }

  // Simulate realistic inter-round pause (agents think between rounds)
  sleep(Math.random() * 0.5 + 0.1);
}

// ── Summary override for cleaner CI output ───────────────────────────────────
export function handleSummary(data) {
  const p50  = data.metrics.negotiation_round_latency?.values?.["p(50)"]  ?? 0;
  const p95  = data.metrics.negotiation_round_latency?.values?.["p(95)"]  ?? 0;
  const p99  = data.metrics.negotiation_round_latency?.values?.["p(99)"]  ?? 0;
  const stP99 = data.metrics.stability_check_latency?.values?.["p(99)"]  ?? 0;
  const failures = data.metrics.integrity_failures?.values?.count ?? 0;
  const successRate = (data.metrics.session_success_rate?.values?.rate ?? 0) * 100;
  const errRate = (data.metrics.http_req_failed?.values?.rate ?? 0) * 100;

  const report = [
    "# GridNexus Load Test Report",
    "",
    "## Negotiation-Round Latency",
    `| Metric | Value |`,
    `|--------|-------|`,
    `| p50  | ${p50.toFixed(1)} ms |`,
    `| p95  | ${p95.toFixed(1)} ms |`,
    `| p99  | ${p99.toFixed(1)} ms |`,
    `| Target (p99 < 2000 ms) | ${p99 < 2000 ? "✅ PASS" : "❌ FAIL"} |`,
    "",
    "## Stability Check Latency",
    `| p99 | ${stP99.toFixed(1)} ms | ${stP99 < 2000 ? "✅ PASS" : "❌ FAIL"} |`,
    "",
    "## Reliability",
    `| Metric | Value |`,
    `|--------|-------|`,
    `| HTTP error rate | ${errRate.toFixed(2)}% |`,
    `| Session success rate | ${successRate.toFixed(1)}% |`,
    `| Integrity failures | ${failures} |`,
    "",
    `Run at: ${new Date().toISOString()}`,
    `VUs: 200 concurrent, 60s sustained`,
  ].join("\n");

  return {
    "load-tests/results/LOAD_TEST_REPORT.md": report,
    stdout: "\n" + report + "\n",
  };
}
